File: gc-starter/redis-sping-boot-starter/src/main/java/com/gc/starter/redis/service/impl/RedisServiceImpl.java
package com.gc.starter.redis.service.impl;

import com.gc.starter.redis.service.RedisService;
import org.springframework.data.redis.connection.RedisStringCommands;
import org.springframework.data.redis.core.RedisCallback;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.core.types.Expiration;
import org.springframework.data.redis.serializer.RedisSerializer;
import org.springframework.lang.NonNull;
import org.springframework.lang.Nullable;

import java.util.Collection;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * redis服务层
 * @author shizhongming
 * 2020/1/17 8:47 下午
 */
public class RedisServiceImpl implements RedisService {

    private RedisTemplate<Object, Object> redisTemplate;

    public RedisServiceImpl(RedisTemplate<Object, Object> redisTemplate) {
        this.redisTemplate = redisTemplate;
    }

    @Override
    public void matchDelete(@NonNull Object prefixKey) {
        // todo: 待开发
    }

    @Override
    public void put(@NonNull Object key, @NonNull Object value) {
        redisTemplate.opsForValue().set(key, value);

    }

    @Override
    public void put(@NonNull Object key, @NonNull Object value, long timeout) {
        redisTemplate.opsForValue().set(key, value, timeout, TimeUnit.SECONDS);
    }

    @Override
    public void put(@NonNull Object key, @NonNull Object value, @NonNull Date expireTime) {
        this.put(key, value);
        redisTemplate.expireAt(key, expireTime);
    }

    @Override
    public void batchPut(@NonNull Map<Object, Object> keyValues) {
        this.redisTemplate.opsForValue().multiSet(keyValues);
    }

    @Override
    public void batchPut(@NonNull Map<Object, Object> keyValues, long timeout) {
        this.redisTemplate.executePipelined((RedisCallback<Object>) connection -> {
            RedisSerializer keySerializer = this.redisTemplate.getKeySerializer();
            RedisSerializer valueSerializer = this.redisTemplate.getValueSerializer();
            keyValues.forEach((key, value) -> connection.set(keySerializer.serialize(key), valueSerializer.serialize(value), Expiration.seconds(timeout),  RedisStringCommands.SetOption.UPSERT));
            return null;
        });
    }

    /**
     * 设置key的过期时间
     * @param key key
     * @param timeout 过期时间
     */
    @Override
    public void expire(@NonNull Object key, long timeout) {
        this.redisTemplate.expire(key, timeout, TimeUnit.SECONDS);
    }

    /**
     * 批量设置key的过期时间
     * @param keys key
     * @param timeout 过期时间
     */
    @Override
    public void batchExpire(@NonNull Collection<Object> keys, long timeout) {
        keys.forEach(key -> this.expire(key, timeout));
    }

    @Override
    public void batchPut(@NonNull Map<Object, Object> keyValues, @NonNull Date expireTime) {
        this.redisTemplate.executePipelined((RedisCallback<Object>) connection -> {
            RedisSerializer keySerializer = this.redisTemplate.getKeySerializer();
            RedisSerializer valueSerializer = this.redisTemplate.getValueSerializer();
            keyValues.forEach((key, value) -> connection.set(keySerializer.serialize(key), valueSerializer.serialize(value), Expiration.milliseconds(expireTime.getTime() - System.currentTimeMillis()),  RedisStringCommands.SetOption.UPSERT));
            return null;
        });
    }

    @Override
    public @Nullable Object get(@NonNull Object key) {
        return this.redisTemplate.opsForValue().get(key);
    }

    @Override
    public <T> T get(@NonNull Object key, @NonNull Class<T> clazz) {
        return (T) this.get(key);
    }

    @Override
    public @Nullable List<Object> batchGet(@NonNull Collection<Object> keys) {
        return this.redisTemplate.opsForValue().multiGet(keys);
    }

    @Override
    public @Nullable <T> List<T> batchGet(@NonNull Collection<Object> keys, @NonNull Class<T> clazz) {
        return (List<T>) this.redisTemplate.opsForValue().multiGet(keys);
    }

    @Override
    public void delete(@NonNull Object key) {
        this.redisTemplate.delete(key);
    }

    @Override
    public void batchDelete(@NonNull List<Object> keys) {
        this.redisTemplate.delete(keys);
    }
}

File: smart-cache/redis-spring-boot-starter/src/main/kotlin/com/smart/starter/redis/service/impl/RedisServiceImpl.kt
package com.smart.starter.redis.service.impl

import com.smart.starter.redis.service.RedisService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.redis.connection.RedisStringCommands
import org.springframework.data.redis.core.RedisTemplate
import org.springframework.data.redis.core.types.Expiration
import org.springframework.data.redis.serializer.RedisSerializer
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * redis缓存服务
 * @author ming
 * 2019/6/22 下午3:28
 */
class RedisServiceImpl : RedisService {

    @Autowired
    private lateinit var redisTemplate: RedisTemplate<String, Any>

    /**
     * 写入缓存
     * @param key 缓存的key
     * @param value 缓存的value
     */
    override fun put(key: String, value: Any) {
        val valueOperations = redisTemplate.opsForValue()
        valueOperations.set(key, value)
    }


    /**
     * 写入缓存，并设置有效时间
     * @param key 缓存的key
     * @param value 缓存的value
     * @param timeout 缓存的有效时间（单位：秒）
     */
    override fun put(key: String, value: Any, timeout: Long) {
        val valueOperations = redisTemplate.opsForValue()
        valueOperations.set(key, value, timeout, TimeUnit.SECONDS)
    }

    /**
     * 写入缓存，并设置过期时间
     * @param key 缓存的key
     * @param value 缓存的value
     * @param expireTime 缓存过期时间
     */
    override fun put(key: String, value: Any, expireTime: Date) {
        this.put(key, value)
        redisTemplate.expireAt(key, expireTime)
    }

    /**
     * 批量写入缓存
     * @param keyValues 缓存键值对
     */
    override fun batchPut(keyValues: Map<String, Any>) {
        this.redisTemplate.opsForValue().multiSet(keyValues)
    }

    /**
     * 批量写入缓存,并设置有效时间
     * @param keyValues 缓存键值对
     * @param timeout 缓存有效时间（单位：秒）
     */
    override fun batchPut(keyValues: Map<String, Any>, timeout: Long) {
        redisTemplate.executePipelined { connection ->
            val keySerializer = redisTemplate.keySerializer as RedisSerializer<String>
            val valueSerializer = redisTemplate.valueSerializer as RedisSerializer<Any>
            for ((key, value) in keyValues) {
                connection.set(keySerializer.serialize(key)!!, valueSerializer.serialize(value)!!,
                        Expiration.seconds(timeout), RedisStringCommands.SetOption.UPSERT)

            }
        }
    }

    /**
     * 批量写入缓存,并设置过期时间
     * @param keyValues 缓存键值对
     * @param expireTime 缓存过期时间
     */
    override fun batchPut(keyValues: Map<String, Any>, expireTime: Date) {
        redisTemplate.executePipelined { connection ->
            val keySerializer = redisTemplate.keySerializer as RedisSerializer<String>
            val valueSerializer = redisTemplate.valueSerializer as RedisSerializer<Any>
            for ((key, value) in keyValues) {
                connection.set(keySerializer.serialize(key)!!, valueSerializer.serialize(value)!!,
                        Expiration.milliseconds(expireTime.getTime() - System.currentTimeMillis()),
                        RedisStringCommands.SetOption.UPSERT)
            }
        }
    }

    /**
     * 读取缓存
     * @param key 键
     * @return 值
     */
    override fun get(key: String): Any? {
        return this.redisTemplate.opsForValue().get(key)
    }

    /**
     * 读取缓存
     * @param key 键
     * @param clazz 值类型
     * @param <T> 值类型
     * @return 值
    </T> */
    override fun <T> get(key: String, clazz: Class<T>): T? {
        return this.redisTemplate.opsForValue().get(key) as T?
    }

    /**
     * 批量读取缓存
     * @param keys 键集合
     * @return 值集合
     */
    override fun batchGet(keys: Collection<String>): List<Any>? {
        return this.redisTemplate.opsForValue().multiGet(keys)
    }

    /**
     * 批量量读取缓存
     * @param keys 键集合
     * @param cls 值的类型
     * @param <T> 值的类型
     * @return 值
    </T> */
    override fun <T> batchGet(keys: Collection<String>, cls: Class<T>): List<T>? {
        return this.redisTemplate.opsForValue().multiGet(keys) as List<T>?
    }

    /**
     * 删除缓存
     * @param key 键
     */
    override fun delete(key: String) {
        this.redisTemplate.delete(key)
    }

    /**
     * 批量删除
     * @param keys 键集合
     */
    override fun batchDelete(keys: Collection<String>) {
        this.redisTemplate.delete(keys)
    }


}